File: ntl_tester.py
'''
	#############################
	NetworkTransportLayer Tester v2
	#############################
	
	Written by Henry Harder
	for Paradigm, 2018

	@version 0.0.3
	@version_date: 9 April 2018

	https://paradigm.market/
	https://paradigmfoundation.io/
'''

# import NetworkBroadcastLayer toolkit
from ntl import*
from time import*
import pyperclip as pyp

'''
#########
this is PRIVATE 
#########
'''
app_id = ''
app_key = ''


'''
#########
END PRIVATE DATA 
#########
'''


order = {'data':{'message':'not really, hello'}}

# generate a random test keypair
# in reality, real keys would be used
alice = KeypairGenerator()

# create instance of NTL
broadcaster = NetworkTransportLayer(app_id, app_key)
reader = OrderBookReader()

# create a transaction
test_tx = broadcaster.make_singlesign_order(
	order, {'timestamp':time()}, alice.get_public_key()) 

signed_tx = broadcaster.sign_order(test_tx, alice.get_private_key())
pyp.copy(signed_tx['id'])
sent_order = broadcaster.send_order(signed_tx)
print('sent. timing.')
delay = 0
while (len(reader.get_full_asset(signed_tx['id']))==2):
	delay += 1
print(time()-reader.get_full_asset(signed_tx['id'])['metadata']['timestamp'])

File: ntl-js/ntl_tester.js
 /*
	NTL Tester Program (v0.1.2b)
	NetworkTransportLayerJS 

	@version-date: 17 April 2018
	@author: Henry R Harder

	Written for Paradigm, 2018

	Description:
		This class provides complete functionality for demonstrating
		creation and broadcast of a sample order to a BigchainDB cluster
		specified in the constructor. You must fill in the 'app_id' and
		'app_key' fields with actual BigchainDB credentials if you wish
		to use the testnet. 
*/

 class NetworkTransportLayer{
 	/*
	This is the main NTL class that communicates with BigchainDB
	and facilitates the creation, signing, and broadcast of orders.
	It must be supplied with credentials for accessing a BigchainDB
	cluster.
	*/

	constructor(id, key, rootUrl){
		this.driver = require('bigchaindb-driver');
		this.connection = new this.driver.Connection(rootUrl, {
			app_id : id,
			app_key : key
		});
	}

	makeOrder(data, metadata, publicKey){
		const newOrder = this.driver.Transaction.makeCreateTransaction(
			data, metadata,
			[this.driver.Transaction.makeOutput(
				this.driver.Transaction.makeEd25519Condition(publicKey))
			], publicKey);

		return newOrder;
	}

	signOrder(unsignedOrder, privateKey){
		const signedOrder = this.driver.Transaction.signTransaction(
			unsignedOrder, privateKey);

		return signedOrder;
	}

	sendOrder(signedOrder){
		this.connection.postTransactionCommit(signedOrder);
		return signedOrder.id;
	}
}

class KeyPairGenerator{
	/*
	This class creates a public/private key pair that can be
	used for testing purposes, and provides methods to access
	the public and private keys. Should not be used for real 
	key generation, meerly for testing purposes. 
	*/

	constructor(){
		this.driver = require('bigchaindb-driver');
		this.pair = new this.driver.Ed25519Keypair();
	}

	getPublicKey(){
		return this.pair.publicKey;
	}

	getPrivateKey(){
		return this.pair.privateKey;
	}
}

class OrderMaker{
	/*
	This class takes a list of uints, a list of hex
	addresses, and a maker address, and provides 
	methods for returning a formatted order and order
	metadata ready for broadcast to BigchainDB
	*/

	constructor(intList, addrList, maker){
		this.orderData = {
			'order':{
				"maker" : [],
				"timestamp" : Math.floor(Date.now()/1000),
				"fields" : {
					"intList" : intList,
					"addrList" : addrList
				}
			}
		};
		this.orderMetaData = {
			"metadata":{
				"filled" : false,
				"valid" : true,
				"updated" : Math.floor(Date.now()/1000)
			}
		};
	}

	getOrderData(){
		return this.orderData;
	}

	getOrderMetadata(){
		return this.orderMetaData;
	}
}

function submitOrder(){
	testntl = new NetworkTransportLayer(
	'', // insert your credentials here to test
	'', // insert your credentials here to test
	'https://test.bigchaindb.com/api/v1/',
	 ); 

	var intList = [];
	var addrList = [];

	intList.push(document.getElementById("uint1").value);
	intList.push(document.getElementById("uint2").value);
	intList.push(document.getElementById("uint3").value);

	addrList.push(document.getElementById("addr1").value);
	addrList.push(document.getElementById("addr2").value);
	addrList.push(document.getElementById("addr3").value);

	alice = new KeyPairGenerator();
	order = new OrderMaker(
		intList,
		addrList,
		alice.getPublicKey());

	newOrder = testntl.makeOrder(order.getOrderData(), order.getOrderMetadata(), alice.getPublicKey());

	signedOrder = testntl.signOrder(newOrder, alice.getPrivateKey());

	orderID = testntl.sendOrder(signedOrder);

	document.write(orderID);
}